File: CoUML-app/Models/Component.cs
using System;
using System.Collections.Generic;
using System.Linq;
using CoUML_app.Controllers.Generators;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace CoUML_app.Models
{
	public abstract class Component : UmlElement{
		public string name { get;set; }

		override public void GenerateCode(ISourceCodeGenerator codeGenerator)
		{
			codeGenerator.Parse(this);
		}

		override protected void ApplyLocally(ChangeRecord change)
		{
			switch(change.action)
			{
				case ActionType.Label: name = (string)change.value; break;
				default: base.ApplyLocally(change);	break;							
			}
		}

	}

	public class Enumeration : Component{
		public ICollection<Enumeral> enums { get;set; }= new RelationalCollection<Enumeral>();

		public Enumeration(): base() { }
		
		override public void Apply(ChangeRecord change, int depth)
		{
			if(AppliesToMe(change.id))
				ApplyLocally(change);
			else
				enums[change.id[++depth]].Apply(change, depth);
		}

		protected override void ApplyLocally(ChangeRecord change)
		{
			switch(change.action)
			{
				case ActionType.Insert: enums.Insert((Enumeral) change.value); break;
				case ActionType.Remove: enums.Remove((string) change.value); break;
				default: base.ApplyLocally(change);	break;							
			}
		}

	}
	
	public class Interface : Component{
		public ICollection<Operation> operations { get;set; }= new RelationalCollection<Operation>();
		public Interface(): base() { }
		
		override public void Apply(ChangeRecord change, int depth)
		{
			if(AppliesToMe(change.id))
				ApplyLocally(change);
			else
				operations[change.id[++depth]].Apply(change, depth);
		}

		protected override void ApplyLocally(ChangeRecord change)
		{
			switch(change.action)
			{
				case ActionType.Insert: operations.Insert((Operation)change.value); break;
				case ActionType.Remove: operations.Remove((string)change.value); break;
				default: base.ApplyLocally(change);	break;							
			}
		}
	}

	public class AbstractClass : Component{
		public ICollection<Operation> operations { get;set; }= new RelationalCollection<Operation>();
		public ICollection<Attribute>  attribute { get;set; }= new RelationalCollection<Attribute>();
		
		public AbstractClass(): base() { }

		override public void Apply(ChangeRecord change, int depth)
		{
			if(AppliesToMe(change.id))
				ApplyLocally(change);
			else
			{
				var next = change.id[++depth];
				ComponentProperty prop = operations[next];
				prop ??= attribute[next];
				prop?.Apply(change, depth);
			}
				
		}

		protected override void ApplyLocally(ChangeRecord change)
		{
			switch(change.action)
			{
				case ActionType.Insert: 
					switch(change.affectedProperty)
					{
						case PropertyType.Attribute: attribute.Insert((Attribute)change.value); break;
						case PropertyType.Operations: operations.Insert((Operation)change.value); break;
						default: break;
					}
					break;
				case ActionType.Remove:
					switch(change.affectedProperty)
					{
						case PropertyType.Attribute: attribute.Remove((string)change.value); break;
						case PropertyType.Operations: operations.Remove((string)change.value); break;
						default: break;
					}
					break;
				case ActionType.Change: 
				default: base.ApplyLocally(change);	break;							
			}
		}
	}

	public class Class : AbstractClass{
		// public ICollection<Operation> operations { get;set; }= new RelationalCollection<Operation>();
		// public ICollection<Attribute> attribute { get;set; }= new RelationalCollection<Attribute>();

		// public Class(string name): base(name) { }
		public Class(): base() { }

		// override public void Apply(ChangeRecord change, int depth)
		// {
		// 	if(AppliesToMe(change.id))
		// 		ApplyLocally(change);
		// 	else
		// 	{
		// 		var next = change.id[++depth];
		// 		ComponentProperty prop = operations[next];
		// 		prop ??= attribute[next];
		// 		prop?.Apply(change, depth);
		// 	}
				
		// }

		// protected override void ApplyLocally(ChangeRecord change)
		// {
		// 	switch(change.action)
		// 	{
		// 		case ActionType.Insert: 
		// 			switch(change.affectedProperty)
		// 			{
		// 				case PropertyType.Attribute: attribute.Insert((Attribute)change.value); break;
		// 				case PropertyType.Operations: operations.Insert((Operation)change.value); break;
		// 			}
		// 			break;
		// 		case ActionType.Remove:
		// 			switch(change.affectedProperty)
		// 			{
		// 				case PropertyType.Attribute: attribute.Remove((string)change.value); break;
		// 				case PropertyType.Operations: operations.Remove((string)change.value); break;
		// 			}
		// 			break;
		// 		case ActionType.Change: 
		// 		default: base.ApplyLocally(change);	break;							
		// 	}
		// }
	}
}

File: CoUML-app/ClientApp/src/models/Component.ts
import { ChangeRecord } from "./ChangeRecord";
import { GeneralCollection, RelationalCollection } from "./Collection";
import { UmlElement } from "./Diagram";
import { Operation, Attribute, ComponentProperty,  ICollection, Enumeral } from "./DiagramModel";


/**
 * Abstraction of Component 
 * Enumerations
 * Interface
 * AbstractClass
 * Class
 */
export abstract class Component extends UmlElement 
{
    public name:string;

    constructor(type: string, name: string)
    {
        super(type);
        this.name = name;
    }

    toUmlNotation(): string {
        return this.name;
    }

    label( newLabel: string)
    {
        this.name = newLabel;
    }
}

/**
 * Enumeration
 */
export class Enumeration extends Component
{
    change(change: ChangeRecord) {
        throw new Error("Method not implemented.");
    }
    insert(element: any) {
        this.enums.insert(element);
    }
    remove(id: string) {
        this.enums.remove(id);
    }
    public enums: ICollection<Enumeral>;

    public constructor(name: string = "EnumerationComponent")
    {
        super("Enumeration", name);
        this.enums  = new RelationalCollection<Enumeral> ([]);
    }

    get(id: string) {
        return this.enums.get(id);
    }
}

/**
 * Interface
 */
export class Interface extends Component
{
    public operations: ICollection<Operation>;

    public constructor(name: string = "InterfaceComponent")
    {
        super("Interface", name);
        this.operations  = new RelationalCollection<Operation> ([]);
    }

    get(id: string) {
        return this.operations.get(id);
    }
    change(change: ChangeRecord) {
        throw new Error("Method not implemented.");
    }
    insert(element: any) {
        this.operations.insert(element);
    }
    remove(id: string) {
        this.operations.remove(id);
    }
}

/**
 * AbstractClass
 */
export class AbstractClass extends Component
{
    change(change: ChangeRecord) {
        throw new Error("Method not implemented.");
    }
    insert(element: any) {
        console.log(element);
        if(element.constructor.name == Operation.name){
            this.operations.insert(element);
        }
        if(element.constructor.name == Attribute.name){
            this.attribute.insert(element);
        }
    }
    remove(id: string) {
       
        let element = this.get(id);
      
        console.log(element);
        if(element.constructor.name == Operation.name){
            this.operations.remove(id);
        }
        if(element.constructor.name == Attribute.name){
            this.attribute.remove(id);
        }
        // return element
    }

    public operations:ICollection<Operation>;
    public attribute:ICollection<Attribute>;

    public constructor(name: string = "AbstractClassComponent")
    {
        super("AbstractClass", name);
        this.operations  = new RelationalCollection<Operation> ([]);
        this.attribute  = new RelationalCollection<Attribute> ([]);
    }


    get(id: any) : Operation | Attribute{
        return this.operations.get(id) || this.attribute.get(id);
    }
}

/**
 * Class
 */
export class Class extends Component
{
    change(change: ChangeRecord) {
        throw new Error("Method not implemented.");
    }
    insert(element: any) {
        
        
        if(element.constructor.name == Operation.name){
            this.operations.insert(element);
        }
        if(element.constructor.name == Attribute.name){
            this.attribute.insert(element);
        }
    }
    remove(id: string) {

        let element = this.get(id);
        console.log(element);
        if(element.constructor.name == Operation.name){
            this.operations.remove(id);
        }
        if(element.constructor.name == Attribute.name){
            this.attribute.remove(id);
        }
        
        // return element
    }

    public operations:ICollection<Operation>;
    public attribute:ICollection<Attribute>;

    public constructor(name: string = "ClassComponent")
    {
        super("Class", name);
        this.operations  = new RelationalCollection<Operation> ([]);
        this.attribute  = new RelationalCollection<Attribute> ([]);
    }


    get(id: any)  : Operation | Attribute{
        return this.operations.get(id) || this.attribute.get(id);
    }
}